pulsar-io/alluxio/src/main/java/org/apache/pulsar/io/alluxio/sink/AlluxioAbstractSink.java

```diff
@@ -0,0 +1,236 @@
+        if (!fileSystem.exists(alluxioDirPath)) {
+            fileSystem.createDirectory(alluxioDirPath);
+        }
+
+        AlluxioURI tmpAlluxioDirPath = new AlluxioURI(tmpFileDirPath);
+        if (!fileSystem.exists(tmpAlluxioDirPath)) {
+            fileSystem.createDirectory(tmpAlluxioDirPath);
+        }
+
+        createFileOptions = CreateFileOptions.defaults();
+
+        recordsNum = 0;
+        recordsToAck = Lists.newArrayList();
+        tmpFilePath = "";
+        alluxioState = AlluxioState.WRITE_STARTED;
+
+        lastRotationTime = System.currentTimeMillis();
+        rotationRecordsNum = alluxioSinkConfig.getRotationRecords();
+        rotationInterval =  alluxioSinkConfig.getRotationInterval();
+    }
+
+    @Override
+    public void write(Record<V> record) {
+        long now = System.currentTimeMillis();
+
+        switch (alluxioState.ordinal()) {
+            case 0:
+                try {
+                    writeToAlluxio(record);
+                    if (!shouldRotate(now)) {
+                        break;
+                    }
+                    alluxioState = AlluxioState.FILE_ROTATED;
+                } catch (AlluxioException | IOException e) {
+                    log.error("Unable to write record to alluxio.", e);
+                    record.fail();
+                }
+            case 1:
+                try {
+                    closeAndCommitTmpFile();
+                    alluxioState = AlluxioState.FILE_COMMITTED;
+                    ackRecords();
+                } catch (AlluxioException | IOException e) {
+                    log.error("Unable to flush records to alluxio.", e);
+                    failRecords();
+                }
+            case 2:
+                alluxioState = AlluxioState.WRITE_STARTED;
+                break;
+            default:
+                log.error("{} is not a valid state when writing record to alluxio temp dir {}.",
+                    alluxioState, tmpFileDirPath);
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        // flush records in the tmpFile when closing sink
+        try {
+            closeAndCommitTmpFile();
+            ackRecords();
+        } catch (AlluxioException | IOException e) {
+            log.error("Unable to flush records to alluxio.", e);
+            failRecords();
+        }
+        deleteTmpFile();
+    }
+
+    private void ackRecords() {
+        recordsToAck.forEach(Record::ack);
+        recordsToAck.clear();
+    }
+
+    private void failRecords() {
+        recordsToAck.forEach(Record::fail);
+        recordsToAck.clear();
+    }
+
+    private void writeToAlluxio(Record<V> record) throws AlluxioException, IOException {
+        KeyValue<K, V> keyValue = extractKeyValue(record);
+        if (fileOutStream == null) {
+            createTmpFile();
+        }
+        fileOutStream.write(keyValue.getValue().toString().getBytes(StandardCharsets.UTF_8));
```

Comment:
I don't think the base class should implement this. It should be sub classes decide how to write the key value pairs. since base class doesn't really know the type of `K` and `V`.

Reply:
changed